crates/sui-bridge/src/sui_syncer.rs

```diff
@@ -1,14 +1,238 @@
+};
+use tokio_retry::strategy::{jitter, ExponentialBackoff};
+use tokio_retry::Retry;
 
+// TODO: use the right package id
+const PACKAGE_ID: ObjectID = SUI_SYSTEM_PACKAGE_ID;
+const SUI_EVENTS_CHANNEL_SIZE: usize = 1000;
 
+/// Map from contract address to their start block.
+pub type SuiTargetModules = HashMap<Identifier, TransactionDigest>;
 
-pub struct EthSyncer<P> {
-    eth_client: Arc<SuiClient>,
-    contract_addresses: EthTargetAddresses,
+pub struct SuiSyncer<C> {
+    sui_client: Arc<SuiClient<C>>,
+    // The last transaction that the syncer has fully processed.
+    // Syncer will resume post this transaction (i.e. exclusive), when it starts.
+    cursors: SuiTargetModules,
+}
+
+impl<C> SuiSyncer<C>
+where
+    C: SuiClientInner + 'static,
+{
+    pub fn new(sui_client: Arc<SuiClient<C>>, cursors: SuiTargetModules) -> Self {
+        Self {
+            sui_client,
+            cursors,
+        }
+    }
+
+    pub async fn run(
+        self,
+        query_interval: Duration,
+    ) -> BridgeResult<(
+        Vec<JoinHandle<()>>,
+        mysten_metrics::metered_channel::Receiver<Vec<SuiEvent>>,
+    )> {
+        let (events_tx, events_rx) = mysten_metrics::metered_channel::channel(
+            SUI_EVENTS_CHANNEL_SIZE,
+            &mysten_metrics::get_metrics()
+                .unwrap()
+                .channels
+                .with_label_values(&["sui_events_queue"]),
+        );
+
+        let mut task_handles = vec![];
+        for (module, cursor) in self.cursors {
+            let events_rx_clone = events_tx.clone();
+            let sui_client_clone = self.sui_client.clone();
+            task_handles.push(spawn_logged_monitored_task!(
+                Self::run_event_listening_task(
+                    module,
+                    cursor,
+                    events_rx_clone,
+                    sui_client_clone,
+                    query_interval
+                )
+            ));
+        }
+        Ok((task_handles, events_rx))
+    }
+
+    async fn run_event_listening_task(
+        // The module where interested events are defined.
+        // Moudle is always of bridge package 0x9.
+        module: Identifier,
+        mut next_cursor: TransactionDigest,
+        events_sender: mysten_metrics::metered_channel::Sender<Vec<SuiEvent>>,
+        sui_client: Arc<SuiClient<C>>,
+        query_interval: Duration,
+    ) {
+        tracing::info!(
+            ?module,
+            ?next_cursor,
+            "Starting sui events listening task from tx_digest {next_cursor}"
+        );
+        let mut interval = time::interval(query_interval);
+        interval.set_missed_tick_behavior(time::MissedTickBehavior::Skip);
+        loop {
+            interval.tick().await;
+            // TODO reconsider panic, should we just log an error and continue the loop?
+            let events = retry_with_max_delay!(
+                sui_client.query_events_by_module(PACKAGE_ID, module.clone(), next_cursor),
```

Comment:
For query_events_by_module, do you know if it's returning events emitted in the module code, or emitted when calling an entry function in the module code?